### ISSUE TITLE
Output probability

### ISSUE THREAD

**Issue by hsellik** [tech-srl/code2seq#36]
Hi @urialon !

I am trying to use this model for bug prediction. Right now I am testing around with binary classification, before moving on to multi-class classification.

However, it would be nice to get a probability for bug/nobug labels that I have defined in JavaExtractor. I understand, that the model (as most seq2seq models) calculates the conditional probability for output tokens.

>  ...the decoder then generates a sequence of output tokens y = (y1, ..., ym) one token at a
time, hence modeling the conditional probability: p (y1, ..., ym|x1, ..., xn).

Basically, this means that I would like to have the (conditional) probability for the very first output token. Tried to look around in [model.py](https://github.com/tech-srl/code2seq/blob/master/model.py), but could not come up with a way to do it.

**Comment by urialon** [comment]
Hi @hsellik ,
The easiest way will be to extract the probabilities from the model.
Paste this snippet before [line 586](https://github.com/tech-srl/code2seq/blob/master/model.py#L586):
```
        logits = outputs.rnn_output # (batch, output_len, classes)
        probs =  tf.nn.softmax(logits, axis=-1)
```
And return the `probs` object as an additional return value.
Then receive the additional returned object in this line: [https://github.com/tech-srl/code2seq/blob/master/model.py#L153](https://github.com/tech-srl/code2seq/blob/master/model.py#L153), 
and finally add it to the list of evaluated tensors in this line:
[https://github.com/tech-srl/code2seq/blob/master/model.py#L187](https://github.com/tech-srl/code2seq/blob/master/model.py#L187). So instead of the current code there:
```
predicted_indices, true_target_strings, top_values = self.sess.run(
                        [self.eval_predicted_indices_op, self.eval_true_target_strings_op, self.eval_topk_values],
                    )
```
Use instead:
```
predicted_indices, true_target_strings, top_values, probs = self.sess.run(
                        [self.eval_predicted_indices_op, self.eval_true_target_strings_op, self.eval_topk_values, self.eval_probs],
                    )
```
where `self.eval_probs` is the object that was returned from `build_test_graph(..)`.
At this point, `probs` will be a numpy array of size `(batch, output_len, num_classes)`, holding the probabilities of the currently evaluated batch.

Since you are addressing classification (i.e., no sequential decoding), you can set `output_len` to `1` [in the config here](https://github.com/tech-srl/code2seq/blob/master/config.py#L22).

I hope it helps. 
Best,
Uri

**Comment by hsellik** [comment]
Perfect, haven't tried it out, but it seems to be the exact thing I need!

Thank you!

**Comment by hsellik** [comment]
I tried it out and it works as expected. The probabilities in num_classes array nicely correspond with indexes of [self.index_to_target](https://github.com/tech-srl/code2seq/blob/master/model.py#L692).

However, I am wondering, is there any special reason why the config.MAX_TARGET_PARTS gets incremented in the code?
[model.py:379](https://github.com/tech-srl/code2seq/blob/master/model.py#L379)
[model.py:476](https://github.com/tech-srl/code2seq/blob/master/model.py#L456)
[model.py:463](https://github.com/tech-srl/code2seq/blob/master/model.py#L463)

**Comment by urialon** [comment]
Great!

`MAX_TARGET_PARTS` is the max number of target "words". In the model, we always add at least one additional padding symbol, such that the model will learn that it has reached the end.
So if the target label is `[bubble, sort]`, the model is actually trained to predict `[bubble, sort, PAD]`.

This starts even earlier, in the reader.
All target sequences that are shorter than the max length are padded, and then we add an additional PAD label (by concatenating the list `[0]`) to all the sequences. In case that one of the sequences is MAX_TARGET_PARTS long, we want to make sure that it will end with a PAD symbol as well:
[https://github.com/tech-srl/code2seq/blob/18b2e8e1e9c4857f39a7f58f32b1872718b2dcec/reader.py#L114](https://github.com/tech-srl/code2seq/blob/18b2e8e1e9c4857f39a7f58f32b1872718b2dcec/reader.py#L114).

Best,
Uri
